### PR TITLE
Bounds checking for Int, and more instances for integer-like types

### LINKED ISSUE #43: Int parser has no bounds checking
The `Int` parser is based on `read`, which behaves very poorly

```haskell
λ> read "436563463456456543453456456543645654654" :: Int
8863761456235732606
```

This leads to unexpected behavior when the parser ought to fail. Test case:

```haskell
λ> execParserPure defaultPrefs (info (parseRecord :: Parser Int) mempty) ["436563463456456543453456456543645654654"]
Success 8863761456235732606
```

### PR DESCRIPTION
**Pull request by chris-martin**
This fixes #43, and also adds instances for `Int8`, `Int16`, `Int32`, `Int64`, `Natural`, `Word8`, `Word16`, `Word32`, and `Word64`.

### CODE REVIEW

File: src/Options/Generic.hs [hunk 1]
```diff
@@ -395,15 +398,43 @@ class ParseField a where
...
 instance ParseField Integer
 instance ParseField Ordering
 instance ParseField ()
 instance ParseField Void
 
+readIntegralBounded :: forall a. (Integral a, Bounded a) => ReadM a
+readIntegralBounded =
+    auto >>= f
+    where
+        f i | i < toInteger (minBound :: a) = empty
```
Comment (Gabriella439) [inline src/Options/Generic.hs]: Could you add more descriptive error messages for these field parse failures?

(Same comment for other `empty`s)

**Review by Gabriella439 — COMMENTED** [review]
One minor comment, but this looks great!

**Comment by Gabriella439** [comment]
Oh, you might also need to guard the `Natural` code behind `CPP` because the integration tests are failing on older versions of `ghc` that precede the introduction of `Natural`s

**Comment by chris-martin** [comment]
Cool, updated.

**Comment by Gabriella439** [comment]
Thanks!  I can upload this to Hackage if you don't have any other pending changes

**Comment by chris-martin** [comment]
Thanks, I don't have anything else.

**Pull request merged**